### ISSUE TITLE
Simplify dictionary encoding with Slice2

### ISSUE THREAD

**Issue by bernardnormier** [icerpc/icerpc-csharp#2670]
A dictionary is really syntactic sugar. We encode/decode a dictionary like a `sequence<(Key, Value)>`, with some restriction on the Key type.

There is not much difference between:
```slice
typealias StringSet = [cs::generic("HashSet")] sequence<string>; 
```
and
```slice
typealias StringInt32Map = dictionary<string, int32>;
```

Seeing dictionaries that way--as a souped-up attribute for sequences--is good, since it simplifies the overall encoding and Slice language.

For Slice1, `dictionary<Key, Value>` is already encoded exactly like `sequence<Pair>` where `Pair = compact struct { k: Key, v: Value } `.

For Slice2 , that's also true, except when the value is an optional type (like `string?`). In this case, we optimize the encoding using a bit sequence at the beginning of the encoded dictionary.

## Proposal
**Always encode `dictionary<Key, Value>` exactly like `sequence<Pair>` where `Pair = compact struct { k: Key, v: Value } `.**
 
in other words, drop the optimization for dictionaries with an optional value type.

If you're sending a huge dictionary with an optional value type, say `dictionary<string, int32?>`, and you need to optimize the encoding, you can simply create your own optimized type. For example:
```slice
compact struct OptimizedStringInt32Map
{
    withValues: dictionary<string, int32>
    withoutValues: sequence<string>
}
```
It's even more compact than our current optimized dictionary-with-optional-value encoding.
